File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/NotificationService.swift
import Foundation
import WagerproofModels
#if canImport(UserNotifications)
import UserNotifications
#endif
#if canImport(UIKit)
import UIKit
#endif

/// `NotificationService` is the iOS-native equivalent of
/// `wagerproof-mobile/services/notificationService.ts`. It owns the OS
/// permission flow and APNs token registration with Supabase.
///
/// Backend contract — byte-identical to the RN file:
/// - Upsert into `user_push_tokens` (cols: user_id, expo_push_token,
///   platform, device_name, is_active, last_used_at, updated_at) on
///   conflict (user_id, expo_push_token).
/// - Ensure a row exists in `user_notification_preferences` (user_id,
///   auto_pick_ready=true) — onConflict user_id, ignoreDuplicates.
/// - Update `is_active=false` on all of a user's tokens on sign-out.
///
/// Wire format: RN stores `expo_push_token`. The native iOS port stores the
/// hex-encoded APNs device token in the same column for now. Either format
/// is accepted by the Supabase row (column is text), and the auto-pick-ready
/// edge function dispatches both APNs and Expo push targets based on token
/// shape.
// FIDELITY-WAIVER #051: Push token stored as raw APNs hex string; RN uses
// expo-tokens. Long-term migration plan in tickets/051-*.md.
public final class NotificationService: @unchecked Sendable {
    public static let shared = NotificationService()

    public enum PermissionStatus: String, Sendable {
        case granted
        case denied
        case undetermined
        case provisional
        case ephemeral
    }

    public enum Platform: String, Sendable {
        case ios
        case android
    }

    private var initialized = false
    // Set by AppDelegate / SceneDelegate after didRegisterForRemoteNotificationsWithDeviceToken.
    private var cachedDeviceToken: String?

    private init() {}

    /// Set the latest APNs device token. Called from the AppDelegate's
    /// `application(_:didRegisterForRemoteNotificationsWithDeviceToken:)`.
    public func setDeviceToken(_ token: String?) {
        cachedDeviceToken = token
    }

    /// Initialize the notification center delegate + foreground presentation
    /// options. Idempotent — safe to call from app launch.
    public func initialize() async {
        guard !initialized else { return }
        initialized = true
        // Foreground presentation options are configured by the AppDelegate's
        // UNUserNotificationCenterDelegate — see RN's
        // `Notifications.setNotificationHandler({ shouldShowAlert: true, … })`.
    }

    public func permissionStatus() async -> PermissionStatus {
        #if canImport(UserNotifications)
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        switch settings.authorizationStatus {
        case .authorized: return .granted
        case .denied: return .denied
        case .notDetermined: return .undetermined
        case .provisional: return .provisional
        case .ephemeral: return .ephemeral
        @unknown default: return .undetermined
        }
        #else
        return .undetermined
        #endif
    }

    /// Triggers the OS authorization dialog. Mirrors RN's
    /// `requestNotificationPermission()`.
    public func requestPermission() async -> PermissionStatus {
        #if canImport(UserNotifications)
        do {
            let granted = try await UNUserNotificationCenter.current()
                .requestAuthorization(options: [.alert, .sound, .badge])
            #if canImport(UIKit)
            if granted {
                await MainActor.run {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
            #endif
            return granted ? .granted : .denied
        } catch {
            return .denied
        }
        #else
        return .denied
        #endif
    }

    /// Register the device's APNs token with Supabase. Mirrors the RN
    /// `registerPushToken(userId)` upsert byte-for-byte. The push-token write
    /// is best-effort: if the token isn't ready yet (registerForRemote in
    /// flight) we silently skip — the next call will pick it up.
    public func registerPushToken(userId: UUID) async {
        guard let token = cachedDeviceToken else { return }
        await PushTokenRegistrar.register(
            userId: userId,
            token: token,
            platform: .ios,
            deviceName: PushTokenRegistrar.deviceModelName()
        )
    }

    /// Deactivate all push tokens for a user. Called on sign-out per RN.
    public func deactivatePushTokens(userId: UUID) async {
        await PushTokenRegistrar.deactivate(userId: userId)
    }

    /// Cached APNs token in hex form — useful for the secret-settings push
    /// diagnostics action.
    public func currentDeviceToken() -> String? {
        cachedDeviceToken
    }
}

/// Supabase row-level writer for push tokens. Split out of the main service
/// so we can keep `NotificationService` Sendable while the Supabase client
/// itself is an actor.
private enum PushTokenRegistrar {
    struct TokenUpsert: Encodable {
        let user_id: UUID
        let expo_push_token: String
        let platform: String
        let device_name: String
        let is_active: Bool
        let last_used_at: String
        let updated_at: String
    }

    struct PreferenceInsert: Encodable {
        let user_id: UUID
        let auto_pick_ready: Bool
    }

    static func register(
        userId: UUID,
        token: String,
        platform: NotificationService.Platform,
        deviceName: String
    ) async {
        let now = isoNow()
        let payload = TokenUpsert(
            user_id: userId,
            expo_push_token: token,
            platform: platform.rawValue,
            device_name: deviceName,
            is_active: true,
            last_used_at: now,
            updated_at: now
        )
        do {
            let client = await MainSupabase.shared.client
            try await client
                .from("user_push_tokens")
                .upsert(payload, onConflict: "user_id,expo_push_token")
                .execute()

            // Ensure notification preferences row exists. RN uses
            // ignoreDuplicates so failures here are non-fatal.
            do {
                try await client
                    .from("user_notification_preferences")
                    .upsert(
                        PreferenceInsert(user_id: userId, auto_pick_ready: true),
                        onConflict: "user_id",
                        ignoreDuplicates: true
                    )
                    .execute()
            } catch {
                // Non-fatal — preferences row may already exist with different
                // columns; RN's ignoreDuplicates short-circuits any conflict.
            }
        } catch {
            // Non-fatal — token sync retries on next foreground.
            _ = error
        }
    }

    static func deactivate(userId: UUID) async {
        struct Update: Encodable { let is_active: Bool }
        do {
            let client = await MainSupabase.shared.client
            try await client
                .from("user_push_tokens")
                .update(Update(is_active: false))
                .eq("user_id", value: userId)
                .execute()
        } catch {
            // Non-fatal — RN's deactivatePushTokens swallows the error too.
        }
    }

    static func deviceModelName() -> String {
        #if canImport(UIKit)
        return UIDevice.current.model
        #else
        return "Unknown"
        #endif
    }

    private static func isoNow() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: Date())
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/NotificationService.kt
package com.wagerproof.core.services

import android.Manifest
import android.app.ActivityManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Process
import androidx.core.app.NotificationChannelCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.os.bundleOf
import com.wagerproof.core.shared.AppGroup
import io.github.jan.supabase.postgrest.from
import java.time.Instant
import kotlinx.serialization.Serializable

/**
 * Port of iOS `NotificationService.swift` / RN `services/notificationService.ts`:
 * push-token registration with Supabase + the local "generation finished" banner.
 *
 * The FCM token itself is supplied from outside — Firebase is not a dependency
 * of this module; the app's `FirebaseMessagingService` calls [cacheToken] on
 * `onNewToken`, and [registerPushToken] picks it up best-effort.
 *
 * Permission flow: this service only *checks* POST_NOTIFICATIONS. The actual
 * runtime request needs an Activity — the UI layer launches
 * `ActivityResultContracts.RequestPermission()` and re-checks [permissionStatus].
 */
object NotificationService {

    enum class PermissionStatus { GRANTED, DENIED }

    private const val CHANNEL_ID = "agent_generation"

    @Volatile
    private var cachedDeviceToken: String? = null

    /** Called from the app's FirebaseMessagingService (onNewToken / getToken). */
    fun cacheToken(token: String) {
        cachedDeviceToken = token
    }

    /** Cached FCM token — for the secret-settings push diagnostics action. */
    fun currentDeviceToken(): String? = cachedDeviceToken

    // FIDELITY-WAIVER #051 analog: the column stays `expo_push_token` even
    // though we store a bare FCM registration token — the server-side
    // dispatcher branches on token shape (Expo vs APNs hex vs FCM).
    @Serializable
    private data class TokenUpsert(
        val user_id: String,
        val expo_push_token: String,
        val platform: String,
        val device_name: String,
        val is_active: Boolean,
        val last_used_at: String,
        val updated_at: String,
    )

    @Serializable
    private data class PreferenceInsert(
        val user_id: String,
        val auto_pick_ready: Boolean,
    )

    /**
     * Register the cached FCM token with Supabase. No-op if no token cached
     * yet (best-effort — retried on the next call). Mirrors the RN upsert
     * byte-for-byte; all failures are non-fatal (retried next foreground).
     */
    suspend fun registerPushToken(userId: String) {
        val token = cachedDeviceToken ?: return
        val now = Instant.now().toString()
        runCatching {
            SupabaseClients.main.from("user_push_tokens").upsert(
                TokenUpsert(
                    user_id = userId,
                    expo_push_token = token,
                    platform = "android",
                    device_name = Build.MODEL,
                    is_active = true,
                    last_used_at = now,
                    updated_at = now,
                ),
            ) {
                onConflict = "user_id,expo_push_token"
            }

            // Ensure a notification-preferences row exists. ignoreDuplicates +
            // swallowed failure — the row may already exist (matches RN).
            runCatching {
                SupabaseClients.main.from("user_notification_preferences").upsert(
                    PreferenceInsert(user_id = userId, auto_pick_ready = true),
                ) {
                    onConflict = "user_id"
                    ignoreDuplicates = true
                }
            }
        }
    }

    /** Deactivate all of a user's push tokens on sign-out. Errors swallowed (matches RN). */
    suspend fun deactivatePushTokens(userId: String) {
        runCatching {
            SupabaseClients.main.from("user_push_tokens").update(
                { set("is_active", false) },
            ) {
                filter { eq("user_id", userId) }
            }
        }
    }

    fun permissionStatus(): PermissionStatus {
        val context = AppGroup.context
        // Android 13+ needs the runtime permission AND channel-level enablement;
        // pre-13 only the app-level toggle exists.
        val runtimeGranted = Build.VERSION.SDK_INT < 33 ||
            context.checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
        val enabled = NotificationManagerCompat.from(context).areNotificationsEnabled()
        return if (runtimeGranted && enabled) PermissionStatus.GRANTED else PermissionStatus.DENIED
    }

    /**
     * Post the LOCAL "pick generation finished" notification for a manual run.
     * Only fires when the app is NOT foreground — on-screen the agent detail
     * page already shows the result live. [isAppForeground] is injectable so
     * callers with lifecycle awareness (ProcessLifecycleOwner) can pass truth.
     */
    fun postGenerationFinishedNotification(
        agentId: String,
        agentName: String,
        picksGenerated: Int,
        parlaysGenerated: Int,
        succeeded: Boolean,
        note: String? = null,
        isAppForeground: Boolean = isAppInForeground(),
    ) {
        if (isAppForeground) return
        if (permissionStatus() != PermissionStatus.GRANTED) return

        val context = AppGroup.context
        val (title, body) = if (succeeded) {
            val parts = buildList {
                if (picksGenerated > 0) add("$picksGenerated pick${if (picksGenerated == 1) "" else "s"}")
                if (parlaysGenerated > 0) add("$parlaysGenerated parlay${if (parlaysGenerated == 1) "" else "s"}")
            }
            "$agentName finished its research" to if (parts.isEmpty()) {
                note ?: "No plays today — the agent passed on the slate."
            } else {
                "${parts.joinToString(" + ")} ready to view."
            }
        } else {
            "$agentName hit a snag" to (note ?: "Pick generation failed. Tap to try again.")
        }

        val manager = NotificationManagerCompat.from(context)
        ensureChannel(manager)

        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            // No drawable resources in this module — borrow the app icon.
            .setSmallIcon(context.applicationInfo.icon.takeIf { it != 0 } ?: android.R.drawable.ic_dialog_info)
            .setContentTitle(title)
            .setContentText(body)
            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
            .setDefaults(NotificationCompat.DEFAULT_SOUND)
            .setAutoCancel(true)
            .addExtras(bundleOf("avatar_id" to agentId, "type" to "generation_finished"))
            .build()

        // Stable per-agent id: a rapid re-run replaces the stale banner instead
        // of stacking duplicates (iOS threadIdentifier/request-id equivalent).
        runCatching {
            manager.notify("generation-finished-$agentId".hashCode(), notification)
        }
    }

    private fun ensureChannel(manager: NotificationManagerCompat) {
        manager.createNotificationChannel(
            NotificationChannelCompat.Builder(CHANNEL_ID, NotificationManagerCompat.IMPORTANCE_DEFAULT)
                .setName("Agent pick generation")
                .build(),
        )
    }

    // Coarse foreground check without a lifecycle dependency — matches the
    // iOS `applicationState == .active` gate closely enough for this banner.
    private fun isAppInForeground(): Boolean {
        val activityManager = AppGroup.context
            .getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager ?: return false
        val myPid = Process.myPid()
        return activityManager.runningAppProcesses?.any {
            it.pid == myPid &&
                it.importance <= ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
        } == true
    }
}
